hack/generator/pkg/codegen/pipeline/unroll_recursive_types.go

```diff
@@ -0,0 +1,188 @@
+// TODO: is simpler and seems to cover all the cases we need.
+// simpleRecursiveTypeFixer removes circular references from types that refer to themselves directly.
+type simpleRecursiveTypeFixer struct {
+	visitor  *astmodel.TypeVisitor
+	newTypes astmodel.Types
+}
+
+type simpleRecursiveTypeFixerContext struct {
+	name         astmodel.TypeName
+	mustUnroll   bool
+	unrolledName astmodel.TypeName
+}
+
+func (c simpleRecursiveTypeFixerContext) WithUnrolledName(name astmodel.TypeName) simpleRecursiveTypeFixerContext {
+	c.unrolledName = name
+	c.mustUnroll = true
+	return c
+}
+
+func newSimpleRecursiveTypeFixer() *simpleRecursiveTypeFixer {
+	result := &simpleRecursiveTypeFixer{
+		newTypes: make(astmodel.Types),
+	}
+
+	visitor := astmodel.TypeVisitorBuilder{
+		VisitObjectType: astmodel.MakeIdentityVisitOfObjectTypeWithPerPropertyContext(result.UnrollObjectTypeProperty),
+		VisitTypeName:   result.UnrollRecursiveReference,
+	}.Build()
+	result.visitor = &visitor
+
+	return result
+}
+
+func (s *simpleRecursiveTypeFixer) UnrollObjectTypeProperty(ot *astmodel.ObjectType, prop *astmodel.PropertyDefinition, ctx interface{}) (interface{}, error) {
+	typedCtx := ctx.(simpleRecursiveTypeFixerContext)
+	name := typedCtx.name
+
+	if !isPropertyMatchingTypeName(name, prop) {
+		return ctx, nil
+	}
+
+	// We know we have at least one property we need to remove. There might be more than 1 direct reference, do a
+	// quick scan for others...
+	toRemove := map[astmodel.PropertyName]struct{}{
+		prop.PropertyName(): {},
+	}
+	for _, p := range ot.Properties().AsSlice() {
+		if _, ok := toRemove[p.PropertyName()]; ok {
+			continue
+		}
+
+		if isPropertyMatchingTypeName(name, p) {
+			toRemove[p.PropertyName()] = struct{}{}
+		}
+	}
+
+	unrolledType := ot
+	for propName := range toRemove {
+		unrolledType = unrolledType.WithoutProperty(propName)
+	}
+	unrolledName := astmodel.MakeTypeName(name.PackageReference, name.Name()+"_Unrolled")
+	unrolledDef := astmodel.MakeTypeDefinition(unrolledName, unrolledType)
+
+	err := s.newTypes.AddAllowDuplicates(unrolledDef)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error adding unrolled type %q", name)
+	}
+
+	return typedCtx.WithUnrolledName(unrolledName), nil
+}
+
+func (s *simpleRecursiveTypeFixer) UnrollRecursiveReference(this *astmodel.TypeVisitor, it astmodel.TypeName, ctx interface{}) (astmodel.Type, error) {
+	typedCtx := ctx.(simpleRecursiveTypeFixerContext)
+
+	if !typedCtx.mustUnroll {
+		return astmodel.IdentityVisitOfTypeName(this, it, ctx)
+	}
+
+	klog.Warningf("%q references itself, removing the self-reference by unrolling to %q", typedCtx.name, typedCtx.unrolledName)
+	return astmodel.IdentityVisitOfTypeName(this, typedCtx.unrolledName, typedCtx)
+}
+
+// asTypeName checks if the type is an astmodel.TypeName or can be unwrapped into an astmodel.TypeName.
+// This differs from astmodel.AsTypeName() in that it unwraps Arrays and Maps too. Use with caution.
```

Comment:
Oh cunning - so this handles `SomeError` with `[]SomeError` or `map[string]SomeError` fields as well? That's spicy!

Reply:
Although this is one of those cases where I think it should be in its own package so another pipeline stage doesn't accidentally use it.

Reply:
I've moved the whole `SimpleRecursiveTypeFixer` to its own package (a subpackage of pipeline)